File: grodbDate/src/nu/mine/mosher/grodb/date/DateRange.java
package nu.mine.mosher.grodb.date;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.Date;

import nu.mine.mosher.core.Immutable;
import nu.mine.mosher.time.Time;

/**
 * TODO
 *
 * @author Chris Mosher
 */
public class DateRange implements Immutable, Serializable, Comparable<DateRange>
{
    /*
	 * YMD always represent Gregorian calendar.
	 * Range of possible dates is given by earliest
	 * thru latest (inclusive).
	 * Date is exact iff earliest.equals(latest).
	 */
	private YMD earliest;
	private YMD latest;

	/**
	 * Indicates what the preferred display calendar is.
	 * true==Julian, false==Gregorian
	 * Note that this indicates only how to display the date(s),
	 * not how they are stored. Dates are always stored using
	 * the Gregorian calendar. Further, it is only a preference,
	 * and therefore the value may be ignored.
	 */
	private boolean julian;

	private boolean circa;

	private transient int hash;
	private transient Time approx;



	static
	{
		assert Immutable.class.isAssignableFrom(YMD.class);
	}

	/**
	 * @param ymd
	 */
	public DateRange(final YMD ymd)
	{
		this(ymd,ymd,false,false);
	}

	/**
	 * @param earliest
	 * @param latest
	 */
	public DateRange(final YMD earliest, final YMD latest)
	{
		this(earliest,latest,false,false);
	}

    /**
     * @param earliest
     * @param latest
     * @param circa
     */
    public DateRange(final YMD earliest, final YMD latest, final boolean circa)
    {
		this(earliest,latest,circa,false);
    }

    /**
     * @param earliest
     * @param latest
     * @param circa
     * @param julian
     */
    public DateRange(final YMD earliest, final YMD latest, final boolean circa, final boolean julian)
    {
		this.earliest = earliest;
        this.latest = latest;
        this.julian = julian;
		this.circa = circa;

		init();
    }

    private void init()
    {
    	if (this.earliest == null)
    	{
    		throw new NullPointerException("earliest date cannot be null");
    	}
		if (this.latest == null)
		{
			throw new NullPointerException("latest date cannot be null");
		}
		if (this.earliest.compareTo(this.latest) > 0)
		{
			throw new IllegalArgumentException("earliest date must be less than or equal to latest date");
		}

		this.approx = calcApprox();
		this.hash = calcHash();
    }



    /**
	 * @return earliest possible date
	 */
	public YMD getEarliest()
	{
		return this.earliest;
	}

    /**
     * @return latest possible date
     */
    public YMD getLatest()
    {
        return this.latest;
    }

	/**
	 * @return if this represents an exact date
	 */
	public boolean isExact()
	{
		return this.earliest.equals(this.latest);
	}

	/**
	 * @return if this date should be show using the Julian calendar
	 * (if so, the caller is responsible for converting it).
	 */
	public boolean isJulian()
	{
		return this.julian;
	}

    /**
     * @return if this is an approximate date
     */
    public boolean isCirca()
    {
        return this.circa;
    }

	/**
	 * @return an approximation of this range
	 */
	public Time getApproxDay()
	{
		return this.approx;
	}



    @Override
	public boolean equals(final Object object)
    {
    	if (!(object instanceof DateRange))
    	{
    		return false;
    	}

		final DateRange that = (DateRange)object;

		return
			this.earliest.equals(that.earliest) &&
			this.latest.equals(that.latest) &&
			this.julian == that.julian &&
			this.circa == that.circa;
    }

    @Override
	public int hashCode()
    {
    	return this.hash;
    }

    @Override
    public String toString()
    {
    	final StringBuffer sb = new StringBuffer();

		if (this.circa)
		{
			sb.append("c. ");
		}

		if (isExact())
    	{
    		sb.append(this.earliest.toString());
    	}
    	else
    	{
    		sb.append("[between ");
    		sb.append(this.earliest.toString());
    		sb.append(" and ");
    		sb.append(this.latest.toString());
    		sb.append("]");
    	}
		return sb.toString();
    }

    public int compareTo(final DateRange that)
    {
        int d = 0;

        if (d == 0)
        {
        	d = this.approx.compareTo(that.approx);
        }
		if (d == 0)
		{
			if (this.circa && !that.circa)
			{
				d = -1;
			}
			if (that.circa && !this.circa)
			{
				d = +1;
			}
		}

        return d;
    }



    private Time calcApprox()
	{
		return new Time(new Date((this.earliest.getApproxTime().asDate().getTime()+this.latest.getApproxTime().asDate().getTime())/2));
	}

	private int calcHash()
	{
		int h = 17;

		h *= 37;
		h += this.earliest.hashCode();
		h *= 37;
		h += this.latest.hashCode();
		h *= 37;
		h += this.julian ? 0 : 1;
		h *= 37;
		h += this.circa ? 0 : 1;

		return h;
	}

    private void writeObject(final ObjectOutputStream s) throws IOException
    {
    	s.defaultWriteObject();
    }

    private void readObject(final ObjectInputStream s) throws IOException, ClassNotFoundException
    {
    	s.defaultReadObject();

    	init();
    }
}

File: grojs/DateRange.js
/**
 * @fileoverview
 * Defines the {@link DateRange} class.
 */

/**
 * @class
 * Represents a range of possible dates.
 * @requires Util
 * @requires YMD
 * 
 * @constructor
 * @param {YMD} earliest 
 * @param {YMD} latest 
 * @return new {@link DateRange}
 * @type DateRange
 */
function DateRange(earliest, latest) {
	Util.verifyType(this,"DateRange");

	this.earliest = earliest;
	if (!this.earliest) {
		throw new Error("missing date");
	}
	this.latest = latest;
	if (!this.latest) {
		this.latest = this.earliest;
	}

	this.approx = this.calcApprox();
}



/**
 * @return earliest possible date
 * @type YMD
 */
DateRange.prototype.getEarliest = function() {
	return this.earliest;
};

/**
 * @return latest possible date
 * @type YMD
 */
DateRange.prototype.getLatest = function() {
	return this.latest;
};

/**
 * @return if this represents an exact date
 * @type Boolean
 */
DateRange.prototype.isExact = function() {
	return YMD.equal(this.earliest,this.latest);
};

/**
 * @return an approximation of this range
 * @type Date
 */
DateRange.prototype.getApproxDay = function() {
	return this.approx;
};

/**
 * @return date string
 * @type String
 */
DateRange.prototype.toString = function() {
	if (this.isExact()) {
		return this.earliest.toString();
	}
	return this.earliest.toString()+"?"+this.latest.toString();
};

/**
 * @private
 * @return an approximation of this range
 * @type Date
 */
DateRange.prototype.calcApprox = function() {
	return new Date((this.earliest.getApproxDate().getMilliseconds()+this.latest.getApproxDate().getMilliseconds())/2);
};

/**
 * Checks two {@link DateRange}s for equality.
 * @param {DateRange} a
 * @param {DateRange} b
 * @return if a and b are equal
 * @type Boolean
 */
DateRange.equal = function(a,b) {
	return YMD.equal(a.getEarliest(),b.getEarliest()) && YMD.equal(a.getLatest(),b.getLatest());
};

/**
 * Compares two {@link DateRange}s, for sorting.
 * @param {DateRange} a
 * @param {DateRange} b
 * @return -1:a<b, 0:a=b, +1:a>b
 * @type Number
 */
DateRange.order = function(a,b) {
	YMD.order(a,b);
};
